backend/drm.c

```diff
@@ -0,0 +1,345 @@
+{
+    *resources = drmModeGetResources(fd);
+    if (*resources == NULL) {
+        log_error("drmModeGetResources failed");
+        return false;
+    }
+    return true;
+}
+
+static drmModeConnectorPtr find_drm_connector(int fd, drmModeResPtr resources)
+{
+    drmModeConnectorPtr connector_ptr = NULL;
+
+    for (int i = 0; i < resources->count_connectors; i++) {
+        connector_ptr = drmModeGetConnector(fd, resources->connectors[i]);
+        if (connector_ptr && connector_ptr->connection == DRM_MODE_CONNECTED) {
+            /* it's connected, let's use this! */
+            break;
+        }
+        drmModeFreeConnector(connector_ptr);
+        connector_ptr = NULL;
+    }
+
+    return connector_ptr;
+}
+
+static bool create_fb(twin_drm_t *tx)
+{
+    /* Create dumb buffer */
+    struct drm_mode_create_dumb create_req = {
+        .width = tx->width,
+        .height = tx->height,
+        .bpp = 32,
+    };
+    if (ioctl(tx->drm_dri_fd, DRM_IOCTL_MODE_CREATE_DUMB, &create_req) < 0) {
+        log_error("Cannot create dumb buffer");
+        return false;
+    }
+    tx->fb_len = create_req.size;
+
+    /* Create framebuffer object for the dumb-buffer */
+    if (drmModeAddFB(tx->drm_dri_fd, tx->width, tx->height, 24, 32,
+                     create_req.pitch, create_req.handle, &tx->fb_id) != 0) {
+        log_error("Cannot create framebubber");
+        munmap(tx->fb_base, tx->fb_len);
+        return false;
+    }
+
+    /* Prepare buffer for memory mapping */
+    struct drm_mode_map_dumb map_req = {
+        .handle = create_req.handle,
+    };
+    if (ioctl(tx->drm_dri_fd, DRM_IOCTL_MODE_MAP_DUMB, &map_req) < 0) {
+        log_error("Cannot map dumb buffer");
+        return false;
+    }
+
+    /* Perform actual memory mapping */
+    tx->fb_base = mmap(NULL, tx->fb_len, PROT_READ | PROT_WRITE, MAP_SHARED,
+                       tx->drm_dri_fd, map_req.offset);
+    if (tx->fb_base == MAP_FAILED) {
+        log_error("Failed to mmap framebuffer");
+        return false;
+    }
+
+    return true;
+}
+
+static bool twin_drm_apply_config(twin_drm_t *tx)
+{
+    /* Retrieve resources */
+    if (!get_resources(tx->drm_dri_fd, &RES(tx)))
+        return false;
+
+    /* Find a connected connector */
+    CONN(tx) = find_drm_connector(tx->drm_dri_fd, RES(tx));
+    if (!CONN(tx))
+        goto bail_conn;
+
+    /* Get the mode information */
+    drmModeModeInfo mode = CONN(tx)->modes[0];
+    tx->width = mode.hdisplay;
+    tx->height = mode.vdisplay;
+
+    /* Create the framebuffer */
+    if (!create_fb(tx))
+        goto bail_fb;
+
+    /* Set the mode on a CRTC */
+    CRTC(tx) = drmModeGetCrtc(tx->drm_dri_fd, RES(tx)->crtcs[0]);
+    if (drmModeSetCrtc(tx->drm_dri_fd, CRTC(tx)->crtc_id, tx->fb_id, 0, 0,
+                       &CONN(tx)->connector_id, 1, &mode) != 0) {
+        log_error("drmModeSetCrtc failed");
+        goto bail_crtc;
+    }
+
+    return true;
+
+bail_crtc:
```

Comment:
I'm confused about this. It should be as below.
bail_fb:
    drmModeRmFB(tx->drm_dri_fd, tx->fb_id);
    munmap(tx->fb_base, tx->fb_len);
bail_crtc:
    drmModeFreeCrtc(CRTC(tx));
    drmModeFreeConnector(CONN(tx));